### PR TITLE
feat: support uri encoding in redirect

### PR DESCRIPTION
**Pull request by Gary-Airwallex**
### What this PR does / why we need it:
Redirect plugin support uri encoding, similar to the implementation in proxy_rewrite.
Fix (part of) https://github.com/apache/apisix/issues/4229.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/plugin/redirect.t [hunk 1]
```diff
@@ -813,3 +813,100 @@ GET /hello
...
+
+
+=== TEST 36: redirect with special characters
+--- request
+GET /hello/with%20space
+--- response_headers
+Location: /hello/with%20space
+--- error_code: 301
+--- no_error_log
+[error]
```
Comment (Firstsawyou) [inline t/plugin/redirect.t:912]: A blank line needs to be added at the end of the file.

File: docs/en/latest/plugins/redirect.md [hunk 2]
```diff
@@ -41,8 +41,9 @@ URI redirect.
...
 | ret_code      | integer | optional    | 302     |  [200, ...]     | Response code                                                                                                                                                                                                                                                                                                                                                                                                                      |
+| encode_uri    | boolean | optional    | false   |       | When set to `true` the uri in `Location` header will be encoded.                                                                                                                                                                                                                         |
 
-Only one of `http_to_https` or `uri` can be specified.
+Only one of `http_to_https`, `uri`, `regex_uri` can be specified.
```
Comment (Firstsawyou) [inline docs/en/latest/plugins/redirect.md:46]:
```suggestion
Only one of `http_to_https`, `uri` or `regex_uri` can be specified.
```
Is this better?

File: docs/zh/latest/plugins/redirect.md [hunk 3]
```diff
@@ -31,8 +31,9 @@ URI 重定向插件。
 | uri           | string  | 可选        |         |            | 可以包含 Nginx 变量的 URI，例如：`/test/index.html`, `$uri/index.html`。你可以通过类似于 `$ {xxx}` 的方式引用变量，以避免产生歧义，例如：`${uri}foo/index.html`。若你需要保留 `$` 字符，那么使用如下格式：`/\$foo/index.html` |
 | regex_uri | array[string] | 可选        |         |                   | 转发到上游的新 `uri` 地址, 使用正则表达式匹配来自客户端的 `uri`，当匹配成功后使用模板替换发送重定向到客户端, 未匹配成功时将客户端请求的 `uri` 转发至上游。`uri` 和 `regex_uri` 不可以同时存在。例如：["^/iresty/(.*)/(.*)/(.*)","/$1-$2-$3"] 第一个元素代表匹配来自客户端请求的 `uri` 正则表达式，第二个元素代表匹配成功后发送重定向到客户端的 `uri` 模板。 |
 | ret_code      | integer | 可选        | 302     | [200, ...] | 请求响应码                                                                                                                                                                                                                    |
+| encode_uri    | boolean | 可选        | false   |       | 当设置为 `true` 时，对返回的 `Location` header进行编码.                                                                                                                                                                               |
 
-`http_to_https` 和 `uri` 两个中只能配置一个。
+`http_to_https`，`uri`，`regex_uri` 三个中只能配置一个。
```
Comment (Firstsawyou) [inline docs/zh/latest/plugins/redirect.md:36]:
```suggestion
`http_to_https`，`uri` 或 `regex_uri` 三个中只能配置一个。
```

File: apisix/plugins/redirect.lua [hunk 4]
```diff
@@ -179,11 +182,23 @@ function _M.rewrite(conf, ctx)
...
+            if n < 1 then
+                return
             end
         end
+
+        if conf.encode_uri then
+            local index = str_find(new_uri, "?")
+            if index then
+                new_uri = core.utils.uri_safe_encode(sub_str(new_uri, 1, index-1)) ..
+                        sub_str(new_uri, index)
```
Comment (Firstsawyou) [inline apisix/plugins/redirect.lua:194]:
```suggestion
                new_uri = core.utils.uri_safe_encode(sub_str(new_uri, 1, index-1)) ..
                          sub_str(new_uri, index)
```
Bad style.

File: docs/en/latest/plugins/redirect.md [hunk 5]
```diff
@@ -41,8 +41,9 @@ URI redirect.
...
 | ret_code      | integer | optional    | 302     |  [200, ...]     | Response code                                                                                                                                                                                                                                                                                                                                                                                                                      |
+| encode_uri    | boolean | optional    | false   |       | When set to `true` the uri in `Location` header will be encoded.                                                                                                                                                                                                                         |
```
Comment (tokers) [inline docs/en/latest/plugins/redirect.md:44]: We may tell people the encoding rule, for instance:

```suggestion
| encode_uri    | boolean | optional    | false   |       | When set to `true` the uri in `Location` header will be encoded as per [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986)|
```

**Review by Firstsawyou — APPROVED** [review]

**Review by starsz — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**